### PR TITLE
[ENH] Lower triple recursion to simple recursion

### PR DESCRIPTION
**Pull request by nouiz**
this fix gh-3341

This inline 3 functions inside only 1. This should also speed up compilation time for big graph.

### CODE REVIEW

File: theano/gof/fg.py [hunk 1]
```diff
@@ -289,16 +293,60 @@ def __remove_clients__(self, r, clients_to_remove,
...
         clients_to_remove
             List of (op, i) pairs such that node.inputs[i] is not r anymore.
+        prune
+            If prune is True, it remove r from this fgraph if it don't
+            have clients left.
+
+        Returns
+        -------
+
+        True if r is still in the fgraph and need to be pruned
```
Comment (abergeron) [inline theano/gof/fg.py]: There shouldn't be a blank line here.  Also the type is mandatory.

**Comment by nouiz** [comment]
I updated the PR.

**Comment by nouiz** [comment]
I fixed the docstring.

**Comment by abergeron** [comment]
The next to last commit passed travis and the last one only modifies a docstring, so I'm merging now.

**Pull request merged**